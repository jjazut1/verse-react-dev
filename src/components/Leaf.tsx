import React from 'react';

// Define the CustomText type for the component
export type CustomText = {
  text: string;
  bold?: boolean;
  italic?: boolean;
  underline?: boolean;
  subscript?: boolean;
  superscript?: boolean;
};

// Define the props for the Leaf component
export interface LeafProps {
  attributes: React.HTMLAttributes<HTMLSpanElement> & {
    'data-slate-leaf'?: true;
  };
  children: React.ReactNode;
  leaf: CustomText;
}

/**
 * Leaf component for rendering formatted text in the Slate editor
 * This component handles the visual representation of text formatting marks
 */
export const Leaf: React.FC<LeafProps> = ({ attributes, children, leaf }) => {
  // Apply each formatting mark by wrapping the children in the appropriate element
  // The order matters for nested elements
  if (leaf.bold) children = <strong>{children}</strong>;
  if (leaf.italic) children = <em>{children}</em>;
  if (leaf.underline) children = <u>{children}</u>;
  if (leaf.subscript) children = <sub>{children}</sub>;
  if (leaf.superscript) children = <sup>{children}</sup>;
  
  // Return the formatted children within a span
  return <span {...attributes}>{children}</span>;
};

export default Leaf; 