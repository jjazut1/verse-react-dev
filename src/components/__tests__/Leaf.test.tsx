import React from 'react';
import renderer from 'react-test-renderer';
import { Leaf, LeafProps } from '../Leaf';

const renderLeaf = (leafProps: Partial<LeafProps>) => {
  const defaultProps: LeafProps = {
    attributes: { 'data-slate-leaf': true },
    children: 'Text',
    leaf: { text: '' }
  };
  
  return renderer.create(
    <Leaf {...defaultProps} {...leafProps} />
  ).toJSON();
};

describe('Leaf Component', () => {
  it('renders plain text correctly', () => {
    const tree = renderLeaf({ leaf: { text: 'Plain text' } });
    expect(tree).toMatchSnapshot();
  });

  it('renders bold text correctly', () => {
    const tree = renderLeaf({ leaf: { text: 'Bold text', bold: true } });
    expect(tree).toMatchSnapshot();
  });

  it('renders italic text correctly', () => {
    const tree = renderLeaf({ leaf: { text: 'Italic text', italic: true } });
    expect(tree).toMatchSnapshot();
  });

  it('renders underline text correctly', () => {
    const tree = renderLeaf({ leaf: { text: 'Underlined text', underline: true } });
    expect(tree).toMatchSnapshot();
  });

  it('renders subscript text correctly', () => {
    const tree = renderLeaf({ leaf: { text: 'Subscript text', subscript: true } });
    expect(tree).toMatchSnapshot();
  });

  it('renders superscript text correctly', () => {
    const tree = renderLeaf({ leaf: { text: 'Superscript text', superscript: true } });
    expect(tree).toMatchSnapshot();
  });

  it('renders combined marks correctly', () => {
    const tree = renderLeaf({
      leaf: { 
        text: 'Combined formatting',
        bold: true, 
        italic: true, 
        underline: true 
      }
    });
    expect(tree).toMatchSnapshot();
  });

  it('renders all formatting combinations correctly', () => {
    const tree = renderLeaf({
      leaf: { 
        text: 'All formatting',
        bold: true, 
        italic: true, 
        underline: true, 
        subscript: true, 
        superscript: true 
      }
    });
    expect(tree).toMatchSnapshot();
  });
  
  // Test to visually inspect the DOM structure
  it('has correct DOM structure for bold text', () => {
    const component = renderer.create(
      <Leaf 
        attributes={{ 'data-slate-leaf': true }} 
        children="Bold text" 
        leaf={{ text: 'Bold text', bold: true }} 
      />
    );
    const tree = component.toJSON();
    console.log('Bold text DOM structure:', JSON.stringify(tree, null, 2));
    
    // Add proper type guards
    expect(tree).not.toBeNull();
    if (tree && !Array.isArray(tree)) {
      expect(tree.type).toBe('span');
      expect(tree.children).not.toBeNull();
      if (tree.children && tree.children.length > 0) {
        const strongElement = tree.children[0] as renderer.ReactTestRendererJSON;
        expect(strongElement.type).toBe('strong');
        expect(strongElement.children).not.toBeNull();
        if (strongElement.children && strongElement.children.length > 0) {
          expect(strongElement.children[0]).toBe('Bold text');
        }
      }
    }
  });
}); 