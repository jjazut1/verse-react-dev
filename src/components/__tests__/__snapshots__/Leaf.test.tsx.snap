// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Leaf Component > renders all formatting combinations correctly 1`] = `
<span
  data-slate-leaf={true}
>
  <sup>
    <sub>
      <u>
        <em>
          <strong>
            Text
          </strong>
        </em>
      </u>
    </sub>
  </sup>
</span>
`;

exports[`Leaf Component > renders bold text correctly 1`] = `
<span
  data-slate-leaf={true}
>
  <strong>
    Text
  </strong>
</span>
`;

exports[`Leaf Component > renders combined marks correctly 1`] = `
<span
  data-slate-leaf={true}
>
  <u>
    <em>
      <strong>
        Text
      </strong>
    </em>
  </u>
</span>
`;

exports[`Leaf Component > renders italic text correctly 1`] = `
<span
  data-slate-leaf={true}
>
  <em>
    Text
  </em>
</span>
`;

exports[`Leaf Component > renders plain text correctly 1`] = `
<span
  data-slate-leaf={true}
>
  Text
</span>
`;

exports[`Leaf Component > renders subscript text correctly 1`] = `
<span
  data-slate-leaf={true}
>
  <sub>
    Text
  </sub>
</span>
`;

exports[`Leaf Component > renders superscript text correctly 1`] = `
<span
  data-slate-leaf={true}
>
  <sup>
    Text
  </sup>
</span>
`;

exports[`Leaf Component > renders underline text correctly 1`] = `
<span
  data-slate-leaf={true}
>
  <u>
    Text
  </u>
</span>
`;
